### PR TITLE
feat: add cpu_serialize weight transfer transport for colocated refit

### PR DESCRIPTION
**Pull request by howard989**
# What does this PR do?

Adds `cpu_serialize` as an alternative weight streaming transport for ZMQ-based model updates. Instead of CUDA IPC handles (which require `CAP_SYS_PTRACE` / `--ipc=host`), this path uses bucket-level pinned DMA + `torch.save`/`torch.load` over multipart ZMQ frames. This makes weight transfer work in restricted container environments without special Docker capabilities.

Key implementation details:
- Ping-pong double buffering with aligned pinned memory for overlapping communication
- ZMQ multipart framing: `[b"cpu_serialize", metadata_pickle, serialized_bytes]`
- Configurable via `model_update_transport="cpu_serialize"` parameter or `NRL_MODEL_UPDATE_TRANSPORT` env var
- ~270ms for 3.4GB on PCIe 4.0 (bucket-level pinned DMA), vs ~2.7s with naive per-tensor `.cpu()`

# Issues

N/A — this is a new transport primitive for the RLix partial-overlap topology (Feature 4).

# Usage

```python
# Direct parameter
policy.stream_weights_via_ipc_zmq(generation, model_update_transport="cpu_serialize")

# Or via environment variable
os.environ["NRL_MODEL_UPDATE_TRANSPORT"] = "cpu_serialize"
```

# Before your PR is "Ready for review"

**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information

## Test Results (Vast.ai, 2x RTX 5070 Ti, nvcr.io/nvidia/nemo-rl:v0.5.0)

Unit tests (test_utils.py): 11/11 passed (95.07s)
Integration tests (test_vllm_generation.py):
  - test_vllm_direct_zmq_weight_update_cpu_serialize: PASSED (493.55s)
  - test_vllm_async_refit_cpu_serialize: PASSED (194.63s)

### Files changed
- `nemo_rl/models/policy/utils.py` — core cpu_serialize send/recv in `stream_weights_via_ipc_zmq_impl()`
- `nemo_rl/models/generation/vllm/vllm_generation.py` — `update_weights_via_ipc_zmq()` delegation
- `nemo_rl/models/generation/vllm/vllm_worker.py` — worker-side `update_weights_via_ipc_zmq()`
- `tests/unit/models/policy/test_utils.py` — parametrized unit tests for both transports
- `tests/unit/models/generation/test_vllm_generation.py` — integration tests (direct ZMQ + async refit)

### CODE REVIEW

**Comment by ZhiyuLi-Nvidia** [comment]
Thanks a lot @howard989 @TianyeGGBond for the contribution. 🙇‍♂️

I verified the speed and accuracy:
image

The accuracy looks pretty good while I am a little bit concerned about the speed and use case:
* Refit time is **~25× slower** when using cpu_serialize.
* Do we have **a specific scenario** that requires cpu_serialize despite that overhead?
  - Even for CPU tensors, would it be faster to stage to GPU and use CUDA IPC instead?

**Comment by taoluo** [comment]
Thank you for taking the time to run the benchmark.
We’re currently working on RFC #2244 and porting several features from ROLL.

For open‑source contributors, we hoped to use vast.ai GPU containers as a lightweight debugging environment. However, vast.ai containers do not provide the ptrace capability required for CUDA IPC (see their [VM feature comparison](https://docs.vast.ai/documentation/instances/virtual-machines#feature-comparison)), so we cannot rely on the standard GPU‑to‑GPU IPC path.

As a result, we’re using this CPU‑serialization workaround while continuing development on the remaining features.

In the broader sense, this enables users to run emerging long‑horizon agentic RL workloads (e.g., swe‑agent) on inexpensive container GPUs from platforms like Runpod or vast.ai. The slowdown introduced by CPU serialization is an acceptable tradeoff for this use case.

For example, with a 30B model, the weight‑sync step takes roughly 90 seconds. This cost is negligible compared to agent rollouts, which can take 30 minutes or more to collect trajectories.

**Comment by ZhiyuLi-Nvidia** [comment]
> Thank you for taking the time to run the benchmark.
> We’re currently working on RFC https://github.com/NVIDIA-NeMo/RL/issues/2244 and porting several features from ROLL.
> 
> For open‑source contributors, we hoped to use vast.ai GPU containers as a lightweight debugging environment. However, vast.ai containers do not provide the ptrace capability required for CUDA IPC (see their [VM feature comparison](https://docs.vast.ai/documentation/instances/virtual-machines#feature-comparison)), so we cannot rely on the standard GPU‑to‑GPU IPC path.
> 
> As a result, we’re using this CPU‑serialization workaround while continuing development on the remaining features.
> 
> In the broader sense, this enables users to run emerging long‑horizon agentic RL workloads (e.g., swe‑agent) on inexpensive container GPUs from platforms like Runpod or vast.ai. The slowdown introduced by CPU serialization is an acceptable tradeoff for this use case.
> 
> For example, with a 30B model, the weight‑sync step takes roughly 90 seconds. This cost is negligible compared to agent rollouts, which can take 30 minutes or more to collect trajectories.


Thanks a lot @taoluo for the explanation. Your use case in [RLix](https://github.com/NVIDIA-NeMo/RL/issues/2244) totally makes sense and our refit duration actually aligns 25s for 8b and projected to 90s for 30b.

Going back to nemo-rl, I'd expect this all the communication will be under the umbrella of nccl eco system.

@terrykong please correct me if I’m wrong: do we anticipate any near-term need to support non-NCCL paths or CPU-only devices? Also, if we do keep this feature, could we host it somewhere other than the core code path to stays clean and easy for maintenance, while still offering it as an alternative option?

**Comment by taoluo** [comment]
Another relevant data point comes from the [Fireworks blog post](https://fireworks.ai/blog/frontier-rl-is-cheaper-than-you-think) on post‑training the Composer‑2 model. They note that:

> “distributing a new checkpoint across globally distributed rollout clusters takes only a few minutes end to end, which bounds the overall off‑policy delay.”

Even with CPU‑serialized weight transfer, the sync cost stays in the “few‑minutes” range for reasonably sized models, and it’s fully amortized by long‑horizon rollouts.

**Comment by ZhiyuLi-Nvidia** [comment]
Hi, @howard989 @taoluo @TianyeGGBond thanks a lot for the contribution. 

Discussed with @terrykong offline, we’re leaning toward not merging this PR right now as the refit time is way too slow Even if it can be overlapped in AsyncRL as @taoluo described, the individual rollout engine still ends up waiting on that wall time.

Your use case in https://github.com/NVIDIA-NeMo/RL/issues/2244 absolutely makes sense for CPU-only machines or environments without CUDA IPC support. For NeMo-RL, though, our near-term direction is to stay within the GPU/NCCL ecosystem.

> Why CPU weight cache + selective sync (not native refit)
> In a multi-pipeline setting, each pipeline's training and inference workers share GPU time with other pipelines. Two problems make NeMo RL's native refit_policy_generation() insufficient:
> 
> Full broadcast causes OOM. refit_policy_generation() broadcasts weights to all inference workers via NCCL collective, requiring the full model staged on GPU while receivers already have their weights loaded. With multiple pipelines sharing GPUs, this is not affordable — each pipeline must sync only to its own shards that actually need updating, not all workers.
> 
> GPU VRAM conflict. Refit requires training weights on GPU when sending (CUDA IPC handles or NCCL broadcast). But when inference workers need to wake up on the same GPUs, both cannot coexist → OOM.
> Solution: CPU weight cache + selective sync.

I think it is partially true as (1) the temporary weight-transfer buffer can be sized safely based on the remaining available memory, and (2) even in memory-constrained scenarios, CPU offloading can still pair with GPU-based transfer as a fast option.

CUDA IPC assumes the co-located GPU has enough memory to hold both the sharded policy weights and the sharded inference copy at the same time. If that assumption holds and we size the temporary weight transfer buffer based on the remaining memory for batched weight transfer appropriately we can likely avoid OOM.

However, if we’re in a tighter memory regime where a co-located GPU can only fit a single sharded model copy, then CPU offload plus **GPU-based weight transfer** is still the better approach. The transfer path on GPU is much faster than CPU-only alternatives. GPU weight transfer has way higher bandwidth, saving unnecessary CPU GPU transfer overhead.

**Comment by taoluo** [comment]
> However, if we’re in a tighter memory regime where a co-located GPU can only fit a single sharded model copy, then CPU offload plus GPU-based weight transfer is still the better approach. The transfer path on GPU is much faster than CPU-only alternatives. GPU weight transfer has way higher bandwidth, saving unnecessary CPU GPU transfer overhead.

Yes, that’s what the RFC intended, though the original wording can be misleading. The primary goal is to minimize the VRAM footprint during refit. On the trainer side, training weights are offloaded to the CPU after each training step, then pipelined and staged into a small GPU buffer before refit using NCCL or CUDA IPC (This PR implements a cpu-serialize fallback as a workaround; it’s fine to keep that path as a development utility for testing on RLix side). The “selective” mechanism sends weights only to the targeted rollout worker, which reduces memory usage on the inference engine.
